File: tryout/lazy-white-falcon.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

var in = bufio.NewReader(os.Stdin)

func getline() string {
	line, _, err := in.ReadLine()
	if err == io.EOF {
		return ""
	}
	if err != nil {
		panic(err)
	}
	return strings.TrimRight(string(line), "\r\n")
}

func getints() []int {
	tokens := strings.Split(getline(), " ")
	ints := make([]int, len(tokens))
	for i, token := range tokens {
		var err error
		ints[i], err = strconv.Atoi(token)
		if err != nil {
			panic(err)
		}
	}
	return ints
}

type LCA struct {
	n             int
	log           int
	parent, depth []int
	parentLift    [][]int
}

func NewLCA(parent, depth []int) *LCA {
	n := len(parent)
	log := 0
	for (1 << log) < n {
		log++
	}
	parentLift := make([][]int, n)
	for i := 0; i < n; i++ {
		parentLift[i] = make([]int, log+1)
		for j := range parentLift[i] {
			parentLift[i][j] = -1
		}
		parentLift[i][0] = parent[i]
	}
	for j := 1; j < log; j++ {
		for i := 0; i < n; i++ {
			if parentLift[i][j-1] != -1 {
				parentLift[i][j] = parentLift[parentLift[i][j-1]][j-1]
			}
		}
	}

	return &LCA{
		n:          n,
		log:        log,
		parent:     parent,
		depth:      depth,
		parentLift: parentLift,
	}
}

func (l *LCA) Find(u, v int) int {
	if l.depth[u] < l.depth[v] {
		u, v = v, u
	}
	for j := l.log; j >= 0; j-- {
		if l.depth[u]-(1<<j) >= l.depth[v] {
			u = l.parentLift[u][j]
		}
	}
	if u == v {
		return u
	}
	for j := l.log; j >= 0; j-- {
		if l.parentLift[u][j] != l.parentLift[v][j] {
			u, v = l.parentLift[u][j], l.parentLift[v][j]
		}
	}
	return l.parent[u]
}

type BIT struct {
	n      int
	values []int
}

func (b *BIT) Update(k, v int) {
	for x := k; x <= b.n; x += (x & -x) {
		b.values[x] += v
	}
}

func (b *BIT) Sum(k int) int {
	sum := 0
	for x := k; x > 0; x -= (x & -x) {
		sum += b.values[x]
	}
	return sum
}

func NewBIT(n int) *BIT {
	return &BIT{
		n:      n,
		values: make([]int, n+1),
	}
}

func main() {
	tokens := getints()
	n := tokens[0]
	q := tokens[1]

	adj := make([][]int, n)
	for i := 0; i < n-1; i++ {
		tokens = getints()
		x := tokens[0]
		y := tokens[1]
		adj[x] = append(adj[x], y)
		adj[y] = append(adj[y], x)
	}

	// DFS for Preorder & LCA
	preL := make([]int, n)
	preR := make([]int, n)
	parent := make([]int, n)
	depth := make([]int, n)

	var dfs func(int, int)
	counter := 0
	dfs = func(v, p int) {
		counter++
		preL[v] = counter
		parent[v] = p
		for _, c := range adj[v] {
			if c == p {
				continue
			}
			depth[c] = depth[v] + 1
			dfs(c, v)
		}
		preR[v] = counter
	}
	depth[0] = 0
	dfs(0, -1)

	lca := NewLCA(parent, depth)
	bit := NewBIT(n + 1)

	value := make([]int, n)
	for iq := 0; iq < q; iq++ {
		tokens = getints()
		t := tokens[0]
		x := tokens[1]
		y := tokens[2]
		if t == 1 {
			diff := y - value[x]
			bit.Update(preL[x], diff)
			bit.Update(preR[x]+1, -diff)
			value[x] += diff
		} else {
			p := lca.Find(x, y)
			ans := bit.Sum(preL[x]) + bit.Sum(preL[y]) - 2*bit.Sum(preL[p]) + value[p]
			fmt.Println(ans)
		}
	}
}

File: tryout/lazy-white-falcon.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
  int n, q;
  scanf("%d%d", &n, &q);
  vector<vector<int>> adj(n);
  for (int i = 0; i < n-1; ++i) {
    int x, y;
    scanf("%d%d", &x, &y);
    adj[x].push_back(y);
    adj[y].push_back(x);
  }

  // DFS for Preorder & LCA
  vector<int> preL(n), preR(n), parent(n), depth(n);
  function<void(int,int)> dfs = [&](int v, int p) {
    static int counter = 0;
    preL[v] = ++counter;
    parent[v] = p;
    for (int c : adj[v]) {
      if (c == p) continue;
      depth[c] = depth[v] + 1;
      dfs(c, v);
    }
    preR[v] = counter;
  };
  depth[0] = 0;
  dfs(0, -1);

  // Precompute binary lifting for LCA
  int log = 0;
  while ((1 << log) < n) log++;
  vector<vector<int>> parentLift(n, vector<int>(log+1, -1));

  for (int i = 0; i < n; ++i) {
    parentLift[i][0] = parent[i];
  }
  for (int j = 1; j < log; ++j) {
    for (int i = 0; i < n; ++i) {
      if (parentLift[i][j-1] != -1) {
        parentLift[i][j] = parentLift[parentLift[i][j-1]][j-1];
      }
    }
  }

  function<int(int,int)> lcaFind = [&](int u, int v) -> int {
    if (depth[u] < depth[v]) swap(u, v);
    for (int j = log; j >= 0; j--) {
      if (depth[u] - (1 << j) >= depth[v]) {
        u = parentLift[u][j];
      }
    }
    if (u == v) return u;
    for (int j = log; j >= 0; --j) {
      if (parentLift[u][j] != parentLift[v][j]) {
        u = parentLift[u][j];
        v = parentLift[v][j];
      }
    }
    return parent[u];
  };

  // BIT/Fenwick Tree
  vector<int> bit(n+1, 0);
  function<void(int,int)> bitUpdate = [&](int k, int v) {
    for (int x = k; x <= n; x += (x & -x)) {
      bit[x] += v;
    }
  };
  function<int(int)> bitSum = [&](int k) -> int {
    int sum = 0;
    for (int x = k; x > 0; x -= (x & -x)) {
      sum += bit[x];
    }
    return sum;
  };

  vector<int> value(n, 0);
  for (int iq = 0; iq < q; ++iq) {
    int t, x, y;
    scanf("%d%d%d", &t, &x, &y);
    if (t == 1) {
      int diff = y - value[x];
      bitUpdate(preL[x], diff);
      bitUpdate(preR[x]+1, -diff);
      value[x] += diff;
    } else {
      int lca = lcaFind(x, y);
      int ans = bitSum(preL[x]) + bitSum(preL[y]) - 2 * bitSum(preL[lca]) + value[lca];
      printf("%d\n", ans);
    }
  }

  return 0;
}